src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -178,27 +179,46 @@ private void InitializeZeroTagPointIfNotInitialized()
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private int LookupAggregatorStore(string[] tagKeys, object[] tagValues, int length)
         {
-            int aggregatorIndex;
-            string[] seqKey = null;
+            var givenTags = new Tags(tagKeys, tagValues);
 
-            // GetOrAdd by TagKeys at 1st Level of 2-level dictionary structure.
-            // Get back a Dictionary of [ Values x Metrics[] ].
-            if (!this.keyValue2MetricAggs.TryGetValue(tagKeys, out var value2metrics))
+            // We only need to sort if there is more than one Tag Key.
+            if (!this.sortedTagsDictionary.TryGetValue(givenTags, out var sortedTags))
             {
                 // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
-                seqKey = new string[length];
-                tagKeys.CopyTo(seqKey, 0);
+                var givenKeys = new string[length];
+                tagKeys.CopyTo(givenKeys, 0);
+
+                var givenValues = new object[length];
+                tagValues.CopyTo(givenValues, 0);
+
+                givenTags = new Tags(givenKeys, givenValues);
+
+                string[] sortedTagKeys;
+                object[] sortedTagValues;
+
+                if (length > 1)
+                {
+                    // Create a new array for the sorted Tag keys.
+                    sortedTagKeys = new string[length];
+                    tagKeys.CopyTo(sortedTagKeys, 0);
+
+                    // Create a new array for the sorted Tag values.
+                    sortedTagValues = new object[length];
+                    tagValues.CopyTo(sortedTagValues, 0);
 
-                value2metrics = new ConcurrentDictionary<object[], int>(ObjectArrayComparer);
-                if (!this.keyValue2MetricAggs.TryAdd(seqKey, value2metrics))
+                    Array.Sort(sortedTagKeys, sortedTagValues);
+                }
+                else
                 {
-                    this.keyValue2MetricAggs.TryGetValue(seqKey, out value2metrics);
+                    sortedTagKeys = givenKeys;
+                    sortedTagValues = givenValues;
                 }
+
+                sortedTags = new Tags(sortedTagKeys, sortedTagValues);
+                this.sortedTagsDictionary.TryAdd(givenTags, sortedTags);
             }
 
-            // GetOrAdd by TagValues at 2st Level of 2-level dictionary structure.
-            // Get back Metrics[].
-            if (!value2metrics.TryGetValue(tagValues, out aggregatorIndex))
+            if (!this.tagsToMetricPointIndexDictionary.TryGetValue(sortedTags, out var aggregatorIndex))
             {
                 aggregatorIndex = this.metricPointIndex;
```

Comment:
I don't think I understand why do we need to have two checks for the boundary, one outside of the lock and one inside the lock. I assume `aggregatorIndex` indicates the current index. I guess there are other modifiers to `tagsToMetricPointIndexDictionary` other than this function which results in having two checks; thus, we want to make sure the current Index is not out of boundary outside of the lock, and the next writable index is valid?

Reply:
"check... take lock..check again" is a standard pattern for maximum performance, and yet correct from multi-threading point.